marketplace_builder/views/pages/use-cases/e-commerce/adding-photos-to-products.liquid

```diff
@@ -0,0 +1,320 @@
+redirect_to: /admin/product/{{ params.form.properties_attributes.product_id }}
+configuration:
+  properties:
+    product_id:
+      validation:
+        presence: true
+  custom_images:
+    image:
+      image:
+        validation:
+          presence:
+            message: 'Upload an image'
+---
+{% comment %}
+  Required params:
+    product_id: string
+{% endcomment %}
+
+{% form, html-class: "my-2", html-data-upload-form: true %}
+  <input type="hidden" name="{{ form_builder.fields.properties.product_id.name }}" value="{{ product_id }}">
+  <div class="form-row align-items-center">
+    <div class="col-auto">
+      <label for="photo-file">Upload a new photo</label>
+    </div>
+    <div class="col-auto">
+      <input
+        class="form-control-file"
+        name="{{ form_builder.fields.custom_images.image.image.name }}"
+        type="file"
+        id="photo-file"
+        accept="image/png, image/jpeg">
+    </div>
+
+    <div class="col-auto">
+      <button type="submit" class="btn btn-primary btn-sm">Upload</button>
+    </div>
+  </div>
+{% endform %}
+
+{% endraw %}
+```
+### Step 2: Add upload script
+
+Upload form should be already working, but is less efficient than it could be. Your file will be uploaded to the application server and then again uploaded to S3 bucket. You can eliminate the intermediate step using S3 credentials provided in `form_builder` object for every `image` or `file` type fields.
+
+Create the upload script first:
+
+#### marketplace_builder/views/partials/forms/upload/s3_script.liquid
+```liquid
+{% raw %}
+{% comment %}
+  Required params:
+    config: s3_upload object from form_builder
+{% endcomment %}
+<script>
+  (function(config){
+    const form = document.querySelector('[data-upload-form]');
+    const input = form.querySelector('input[type="file"]');
+    const submitButton = form.querySelector('[type="submit"]');
+
+    // When S3 upload succeeds, you should:
+    // 1. add hidden input with with the same name attribute as your file input
+    // 2. set it’s value to URL returned from S3
+    // 3. disable file input, so the file is not uploaded again
+    // 4. submit the form
+    function onSuccess(location) {
+      const inputOverride = document.createElement('input');
+      inputOverride.type = 'hidden';
+      inputOverride.name = input.name;
+      inputOverride.value = location;
+      input.disabled = true;
+      form.appendChild(inputOverride);
+      form.submit();
+    }
+
+    // whenever an error occurs you should enable the submit button again
+    // and communicate it to user.
+    function onError(error) {
+      alert('We were unable to upload this file');
+      submitButton.disabled = false;
+      submitButton.innerText = 'Upload';
+      throw error;
+    }
+
+    // handle upload to S3. Make sure you send all of the metadata provided in s3_upload key
+    function s3Upload(file, uploadUrl, metadata) {
+      const request = new XMLHttpRequest();
```

Comment:
i would use fetch for code readability and future proof

Reply:
`fetch` is not available without polyfill in IE11. We need to support it still, WeDo stats for example are reporting almost 10% of users in New Zealand use it. 

XHR on the other hand is not going anywhere. I am sticking with it

Reply:
We shouldn't mix things up - % of users using some particular project has nothing to do with what we want to show - and we want to show how to use the platform in the simplest possible way - doesnt matter how dev sends the request, as long as it gets to the server.

Reply:
Yes. And in this example I am building I choose to support IE11 because at the time of writing it is still a platform that needs to be supported. 

It’s already bad that we have to use JS in the first place to upload a file, I will not make it even worse by using an API that breaks the simplest form for 10% of users.

I will not change it.

Reply:
Showing best practices and explaining why we do something the way we do was an explicit request from community members.

I think adding an explanation could solve this issue. We show one approach, describe why we chose that in this example, and mention the other approach with a link for further reading. What do you think?